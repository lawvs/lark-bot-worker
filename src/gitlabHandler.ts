import type { LarkRobot } from "./lark/robot";
import type { WebhookEvents } from "gitlab-event-types";

/**
 * See https://open.feishu.cn/tool/cardbuilder?from=howtoguide
 */
export const makeInteractiveCard = ({
  title,
  content,
  url,
  at,
  atAll,
}: {
  title: string;
  content: string;
  url?: string;
  /**
   * Custom bot only supports `@users` using open_id;
   */
  at?: string[];
  atAll?: boolean;
}) =>
  ({
    config: {
      wide_screen_mode: true,
    },

    header: {
      template: "purple",
      title: {
        tag: "plain_text",
        content: title,
      },
    },

    elements: [
      {
        tag: "markdown",
        content,
      },

      at && {
        tag: "div",
        text: {
          content: at.map((email) => `<at email=${email}></at>`).join(" "),
          tag: "lark_md",
        },
      },

      atAll && {
        tag: "div",
        text: {
          content: "<at id=all></at>",
          tag: "lark_md",
        },
      },

      url && {
        actions: [
          {
            tag: "button",
            text: {
              content: "立即查看",
              tag: "plain_text",
            },
            type: "primary",
            url,
          },
        ],
        tag: "action",
      },
    ].filter(Boolean),
  }) as const;

export const handleGitlabWebhook = async (
  event: WebhookEvents,
  robot: LarkRobot,
) => {
  if ("object_kind" in event && event.object_kind === "note") {
    const card = makeInteractiveCard({
      title: `${event.project.name} 有新的评论`,
      content: [
        event.merge_request?.title ? `**${event.merge_request.title}**` : "",
        event.object_attributes.note,
      ]
        .filter(Boolean)
        .join("\n"),
      url: event.object_attributes.url,
    });

    const resp = await robot.send({
      msg_type: "interactive",
      card,
    });
    return resp;
  }
  console.error("Unknown event:", event);
};
