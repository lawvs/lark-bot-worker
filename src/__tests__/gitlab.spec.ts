import { describe, expect, vi, test } from "vitest";
import { handleGitlabWebhook } from "../gitlabHandler";
import type { LarkRobot } from "../lark/robot";
import note from "./fixtures/note.json";

describe("gitlab", () => {
  test("should send data match snapshot", async () => {
    const mockRobot: LarkRobot = {
      send: vi.fn((...args) => {
        expect(args).toMatchSnapshot();
        return Promise.resolve({ code: 1, msg: "" });
      }),
    };
    await handleGitlabWebhook(note as any, mockRobot);
    expect(mockRobot.send).toBeCalledTimes(1);
  });
});
