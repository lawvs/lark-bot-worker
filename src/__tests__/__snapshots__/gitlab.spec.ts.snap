// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`gitlab > should send data match snapshot 1`] = `
[
  {
    "card": {
      "config": {
        "wide_screen_mode": true,
      },
      "elements": [
        {
          "content": "**Feat/build report select**
added 1 commit",
          "tag": "markdown",
        },
        {
          "actions": [
            {
              "tag": "button",
              "text": {
                "content": "立即查看",
                "tag": "plain_text",
              },
              "type": "primary",
              "url": "https://gitlab.com/username/projectA/-/merge_requests/1#note_2207893",
            },
          ],
          "tag": "action",
        },
      ],
      "header": {
        "template": "purple",
        "title": {
          "content": "projectA 有新的评论",
          "tag": "plain_text",
        },
      },
    },
    "msg_type": "interactive",
  },
]
`;
